soccer/motion/MotionControlNode.cpp

```diff
@@ -1,28 +1,30 @@
 #include "MotionControlNode.hpp"
 #include "Robot.hpp"
 
-MotionControlNode::MotionControlNode(Context* context)
-    : _context(context), _controllers(Num_Shells, std::nullopt) {
+MotionControlNode::MotionControlNode(Context* context) : _context(context) {
     _controllers.reserve(Num_Shells);
-    for (OurRobot* robot : context->state.self) {
-        _controllers[robot->shell()] = MotionControl(context, robot);
+    for (int i = 0; i < Num_Shells; i++) {
+        _controllers.emplace_back(MotionControl(context, i));
     }
 }
 
 void MotionControlNode::run() {
-    bool force_stop = _context->game_state.state == GameState::State::Halt;
-    for (auto& maybe_controller : _controllers) {
-        if (!maybe_controller) {
-            continue;
-        }
-
-        auto controller = maybe_controller.value();
+    runMotion(_context->world_state, _context->game_state, _context->paths,
+              &_context->motion_setpoints);
+}
 
+void MotionControlNode::runMotion(
+    const WorldState& world_state, const GameState& game_state,
+    const std::array<Planning::AngleFunctionPath, Num_Shells>& paths,
+    std::array<MotionSetpoint, Num_Shells>* setpoints) {
+    bool force_stop = game_state.state == GameState::State::Halt;
+    for (int i = 0; i < Num_Shells; i++) {
+        MotionControl& controller = _controllers[i];
+        MotionSetpoint* setpoint = &(*setpoints)[i];
```

Comment:
This looks a little jank from a syntax pov, but there's not really a way to fix it :/

Reply:
I mean, it's not really that weird. It's "address of the value at index i in the array pointed to by setpoints", which is a pretty reasonable thing to express. Either way not much we can do about it.